### PR TITLE
Code review for Raimondi and Martignago

### PR DESCRIPTION
**Pull request by CremaLuca**
## Premise
This pull request might take some time to review if you have no background about it, so there is a [Trello card](https://trello.com/c/vNEnJR4b) (version 2.0 - network implementation) that might help you understand correctly what's going on if you want to go faster and have a deeper look at the steps I took to arrive here.

## Code
What you're going to see will be 6 interfaces with 0 to 2 methods each (very easy and quick to read) that define a behaviour of a network of "Events": an Event is _something_ happening in a place (while a TimedEvent is still _something_ happening in a place at a time/before a time/after a time).

Together with the interfaces, there will be three implementations of three of those interfaces: `GenericNetworkEvent`, `GenericTimedEvent `and `GenericEventNetwork`; the first two are just variable holders, the third one will be a little more complex but I'm sure you'll understand it right away.

There is still one TODO and it's about a not-so-easy algorithm to understand which positions are inside a given range in meters, I already wrote the first version of it but it looks terrible and it would need a pull request on its own to be reviewed.

## Before you start

1. Please don't look at `GPSPosition` as it's been implemented by Giorgia Bortoletti and I needed it in order to work properly with it in the manager

2. **TESTING IS STILL ON ITS WAY**, it's not easy to mock a location or a NetworkManager from the network library/dictionary.

Have fun.

### CODE REVIEW

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+/**
+ * @author Luca Crema
+ * @since 26/12/2019
+ */
+@RunWith(MockitoJUnitRunner.class)
```
Comment (CremaLuca) [inline app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java]: Codacy Issue found: [Class should define a constructor.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- app/src/main/java/com/eis/geoCalendar/app/network/GenericEventNetwork.java  2
- app/src/main/java/com/eis/geoCalendar/app/network/GenericNetworkTimedEvent.java  1
- app/src/main/java/com/eis/geoCalendar/app/network/GenericNetworkEvent.java  1
- app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777&bid=15825100)

File: app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java [hunk 2]
```diff
@@ -1,36 +1,52 @@
...
-        //TODO: There can be more constructors
+    private Location position;
+
+    public GPSPosition(double latitude, double longitude) {
+        position = new Location("GPSPosition");
+        position.setLatitude(latitude);
+        position.setLongitude(longitude);
+    }
+
+    public GPSPosition(Location position){
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java]: Codacy Issue found: ['{' is not preceded with whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    private GenericEventNetwork<NetworkEvent<String, NetworkEventUser<Peer<String>>>, NetworkEventUser<Peer<String>>> testNetwork;
+    @Mock
+    private Location mockedLocation1;
+    @Mock
+    private EventNetworkManager<NetworkEvent<String, NetworkEventUser<Peer<String>>>, NetworkEventUser<Peer<String>>> eventNetworkManager;
+
+
+    @Before
+    public void setup() {
```
Comment (CremaLuca) [inline app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java]: Codacy Issue found: [You may have misspelled a JUnit framework method (setUp or tearDown)](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * @author Luca Crema
+ * @since 26/12/2019
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class GenericEventNetworkTest {
+
+    private GenericEventNetwork<NetworkEvent<String, NetworkEventUser<Peer<String>>>, NetworkEventUser<Peer<String>>> testNetwork;
+    @Mock
+    private Location mockedLocation1;
```
Comment (CremaLuca) [inline app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java]: Codacy Issue found: [Avoid unused private fields such as 'mockedLocation1'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

File: app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java (hunk 2 above)
Comment (CremaLuca) [reply app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java]: None of my business

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java (hunk 1 above)
Comment (CremaLuca) [reply app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java]: Nah it couldn't, it's a test

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java (hunk 3 above)
Comment (CremaLuca) [reply app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java]: I mean, what's the difference between `setUp` and `setup`

**Review by EdoardoRaimondi — COMMENTED** [review]
### GENERAL COMMENT
First at all, thanks for the documentation. It was complete and easy to read. 
The code is really clean and clear. Lovely generics usages.❤
At the begging I was quite puzzled about that listener usages, but after some code rereadings I convince myself.
Most of the problems I found are already known, as comments and documentation can dimostrate. (testing first at all. With all that (well written) interfaces mock tests are crucial)
I can only hope will be solved. But I have however found some inaccuracies :

### TECHNICAL COMMENTS

### GenericEventNetwork

-line 57 : On a metod called `onGetEvents` the last thing that comes to my mind it's that you will add something. Only if really easy, changing it would be helpful for readable reason.

-line 82 : Is there any way to know specifics reasons? Generics are good only as parameters 😆

-line 94 : What happens if I insert a not reasonable small radius? It will not work properly or simply I will have too positions to handle? What "reasonable" does it mean? From the code I was unable to respond. A little specification would be helpful.

### CONCLUSION

All the little things I liked are a lot and I will spend 10 minutes only writing them. So thanks and very good job!

**Comment by CremaLuca** [comment]
Thank you for your time to review this. I need to clarify a few more things maybe:

`line 57`: The method `onGetEvents` is called when the events are gathered from the network, what it does inside of it is not related to its name because the listener doesn't know what will happen, its name is related to the fact that it's called when _GetEvents_ has happened.

`Line 82`: Oh shoot! I forgot to implement the pass the actual reason in the listener, thanks

`Line 94`: Yeah "reasonable" is not defined because depends on the type of network, and on the approximation used. I could add a description of what parameters "reasonable" depends on.

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java  1
- app/src/main/java/com/eis/geoCalendar/app/network/GenericEventNetwork.java  2
- app/src/main/java/com/eis/geoCalendar/app/network/GenericNetworkEvent.java  1
- app/src/main/java/com/eis/geoCalendar/app/network/GenericNetworkTimedEvent.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777&bid=15825100)

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java (hunk 1 above)
Comment (CremaLuca) [inline app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java]: Codacy Issue found: [Class should define a constructor.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

File: app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java (hunk 2 above)
Comment (CremaLuca) [inline app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java]: Codacy Issue found: ['{' is not preceded with whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- app/src/main/java/com/eis/geoCalendar/app/network/GenericNetworkTimedEvent.java  1
- app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java  1
- app/src/main/java/com/eis/geoCalendar/app/network/GenericEventNetwork.java  3
- app/src/main/java/com/eis/geoCalendar/app/network/GenericNetworkEvent.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777&bid=15825100)

File: app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java (hunk 1 above)
Comment (CremaLuca) [inline app/src/test/java/com/eis/geoCalendar/app/network/GenericEventNetworkTest.java:22]: Codacy Issue found: [Class should define a constructor.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

File: app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java (hunk 2 above)
Comment (CremaLuca) [inline app/src/main/java/com/eis/geoCalendar/gps/GPSPosition.java]: Codacy Issue found: ['{' is not preceded with whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4741777)

**Pull request merged**